libdnf5/utils/iniparser.cpp

```diff
@@ -25,6 +25,49 @@ namespace libdnf5 {
 
 constexpr char DELIMITER = '\n';
 
+namespace {
+
+// Returns the position of the first undoubled character ']'.
+std::size_t find_end_of_section_name(const std::string & str, std::size_t pos) {
+    bool bracket = false;
+    for (std::size_t idx = pos;; ++idx) {
+        const auto ch = str[idx];
+        if (bracket) {
+            if (ch != ']') {
+                return idx - 1;
+            }
+            bracket = false;
+        } else {
+            if (ch == '\0' && ch == '\n' && ch == '\r') {
+                return std::string::npos;
+            }
+            if (ch == ']') {
+                bracket = true;
+            }
+        }
+    }
+}
```

Comment:
The find_end_of_section_name() will exhibit an undefined behavior if "str" does not contain an unescaped ']', '\0', '\n', or '\r'. In that case "idx" will grow indefinitely and the function will overread a heap buffer.
I recommend either limit the for() cycle with str.end() and return std::string::npos when reaching it, or document that the function assumes a '\0'-terminated string.

Reply:
@ppisar 
https://cplusplus.com/reference/string/string/operator[]/
`If pos is equal to the [string length](https://cplusplus.com/string::length), the function returns a reference to the null character that follows the last character in the string (which should not be modified).`

So if the string/memory is not corrupted, `std::string` will always have a null character at the end. It is defined in the standard.

The problem occurs when someone calls the function with the parameter `pos > str.size()` -> skips the terminating null character. But this is not possible from the outside, the function is in an anonymous namespace -> only available locally. But I will modify the code, what if someone makes a mistake in the future.

However, I found another bug in the code. How could I write that. In this line:
`if (ch == '\0' && ch == '\n' && ch == '\r') {`
there should be `if (ch == '\0' || ch == '\n' || ch == '\r') {`

Thanks for the review.

Reply:
I fixed the find_end_of_section_name internal function in IniParser and added unit tests for ini files with incorrect syntax.